### PR TITLE
Feature/ancillaries

### PR DESCRIPTION
**Pull request by vblazenka**
Some notes about what I did:
- Implemented new AncillaryView and AncillaryBridge classes for iOS
- Separate CodePush deployment for ancillaries
- `ancillaries.jsbundle` is generated, ancillaries are separated from `main.jsbundle`
- AncillaryFactory is responsible to return which ancillary to return
- We have only `FastTrack` for now

### CODE REVIEW

File: app/ancillaryFactory/src/appRegistry/AncillaryFactory.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// @flow strict
+
+/* eslint-disable import/order */
```
Comment (tbergq) [inline app/ancillaryFactory/src/appRegistry/AncillaryFactory.js]: please don't do this, fix the import order instead

File: app/ancillaryFactory/src/appRegistry/AncillaryFactory.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import * as React from 'react';
+import CodePush from 'react-native-code-push';
+
+// Config
```
Comment (tbergq) [inline app/ancillaryFactory/src/appRegistry/AncillaryFactory.js]: These comments do not add any necessary information

File: app/ancillaryFactory/src/appRegistry/AncillaryFactory.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+type Props = {|
+  +service: string,
+  +token: string,
+  +bookingId: number,
+|};
+
+class AncillaryFactory extends React.Component<Props> {
+  componentDidMount() {
+    CodePush.sync({
```
Comment (tbergq) [inline app/ancillaryFactory/src/appRegistry/AncillaryFactory.js]: You don't need to do this here, it is already done in `native.js`

File: app/ancillaryFactory/src/components/NoAncillaryMessage.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// @flow strict
+
+import * as React from 'react';
+import { View, Text } from 'react-native'; // eslint-disable-line no-restricted-imports
```
Comment (tbergq) [inline app/ancillaryFactory/src/components/NoAncillaryMessage.js]: We have restricted imports for a reason, please use `Text` from `@kiwicom/mobile-shared` instead

File: app/ancillaryFactory/src/services/createRequester.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+// @flow
+
+import { callApi } from '@kiwicom/mobile-networking';
```
Comment (tbergq) [inline app/ancillaryFactory/src/services/createRequester.js]: we already have `@kiwicom/fetch` for network requests

File: app/fastTrack/src/appRegistry/FastTrackBanner.js [hunk 6]
```diff
@@ -1,28 +1,62 @@
 // @flow strict
 
+/* eslint-disable import/order */
```
Comment (tbergq) [inline app/fastTrack/src/appRegistry/FastTrackBanner.js]: delete this

File: app/fastTrack/src/appRegistry/FastTrackBanner.js [hunk 7]
```diff
@@ -1,28 +1,62 @@
...
+
 import * as React from 'react';
-import {
-  SimpleCard,
-  Text,
-  TextButton,
-  StyleSheet,
-} from '@kiwicom/mobile-shared';
+import { AxiosResponse } from 'axios';
```
Comment (tbergq) [inline app/fastTrack/src/appRegistry/FastTrackBanner.js]: we don't need anymore fetching libraries in the repo

File: app/fastTrack/src/components/FastTrackModal.js [hunk 8]
```diff
@@ -1,46 +1,59 @@
...
+    flex: 1,
+    backgroundColor: 'transparent',
+  },
+  buttonContainer: {
+    paddingHorizontal: Device.isIPhoneX ? 16 : 8,
+    paddingBottom: Device.isIPhoneX ? 32 : 8,
   },
   modal: {
     margin: 0,
+    backgroundColor: 'white',
```
Comment (tbergq) [inline app/fastTrack/src/components/FastTrackModal.js]: should probably be palleteWhite from defaultTokens

File: app/fastTrack/src/appRegistry/FastTrackBanner.js [hunk 9]
```diff
@@ -1,28 +1,62 @@
...
   state = {
     isModalVisible: false,
+    documentUrl: '',
+  };
+
+  componentDidMount() {
+    this.fetchFastTrackDocument();
+  }
+
+  fetchFastTrackDocument = async () => {
```
Comment (tbergq) [inline app/fastTrack/src/appRegistry/FastTrackBanner.js]: Any reason we can not do this in graphql?

File: app/fastTrack/src/components/Services.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
+// @flow stric
+
+import * as React from 'react';
+import { View } from 'react-native';
+import { Text, StyleSheet, Icon } from '@kiwicom/mobile-shared';
+import { Translation } from '@kiwicom/mobile-localization';
+import { defaultTokens } from '@kiwicom/mobile-orbit';
+
+type Service = {
```
Comment (tbergq) [inline app/fastTrack/src/components/Services.js]: Why is this type not exact?

File: ios/RNKiwiMobile/AncillaryConstants.m [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+#else
+  return [CodePush bundleURLForResource:@"ancillaries" withExtension:@"jsbundle" subdirectory:nil bundle:[NSBundle bundleWithIdentifier:@"com.kiwi.RNKiwiMobile"]];
+#endif
+}
+
++ (NSString *)codePushAppVersion {
+  return @"1.0.0";
+}
+
+@end
```
Comment (tbergq) [inline ios/RNKiwiMobile/AncillaryConstants.m]: What is the reason for adding this file?

File: scripts/codePush.js [hunk 12]
```diff
@@ -7,27 +7,43 @@ const childProcess = require('child_process');
...
   exec(
-    `appcenter codepush release-react -a Kiwicom/mobile-android --entry-file ${entryFile} --target-binary-version "${targetBinaryVersion}" -d ${target}`,
+    `appcenter codepush release-react -a ${targetProjectAndroid} --entry-file ${entryFile} --target-binary-version "${targetBinaryVersion}" -d ${target} --bundle-name ${bundleName}`,
   );
 
   exec(
-    `appcenter codepush release-react -a Kiwicom/mobile-ios --entry-file ${entryFile} --target-binary-version "${targetBinaryVersion}" -d ${target}`,
+    `appcenter codepush release-react -a ${targetProjectIOS} --entry-file ${entryFile} --target-binary-version "${targetBinaryVersion}" -d ${target} --bundle-name ${bundleName}`,
   );
 }
```
Comment (tbergq) [inline scripts/codePush.js]: Why these changes?

File: app/ancillaryFactory/src/config/codepushConfig.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
+// @flow strict
+
+import CodePush from 'react-native-code-push';
+import { Platform } from 'react-native';
+
+// TODO: Fill in with Kiwi CodePush keys
```
Comment (RobinCsl) [inline app/ancillaryFactory/src/config/codepushConfig.js]: These should be environment variables that should not be published on GitHub, so you can delete this TODO.
Documentation available here: https://github.com/kiwicom/mobile/blob/master/docs/common/code-push.md

File: app/fastTrack/src/appRegistry/FastTrackBanner.js (hunk 9 above)
Comment (vblazenka) [reply app/fastTrack/src/appRegistry/FastTrackBanner.js]: No GraphQl support for this api.

File: app/fastTrack/src/components/Services.js (hunk 10 above)
Comment (vblazenka) [reply app/fastTrack/src/components/Services.js]: Will fix it :)

File: ios/RNKiwiMobile/AncillaryConstants.m (hunk 11 above)
Comment (vblazenka) [reply ios/RNKiwiMobile/AncillaryConstants.m]: We want to have separate CodePush release process and with this file we manage how to fetch ancillaries bundle and which version of release to use.

File: app/fastTrack/src/appRegistry/FastTrackBanner.js (hunk 9 above)
Comment (RobinCsl) [reply app/fastTrack/src/appRegistry/FastTrackBanner.js]: We do all API calls through our GraphQL proxy, so you would need to create your queries and resolvers there.

File: scripts/codePush.js (hunk 12 above)
Comment (vblazenka) [reply scripts/codePush.js]: We want to have separate CodePush release process for ancillaries.

Comment (tbergq) [reply scripts/codePush.js]: We discussed this before with callstack and native team. We do not want separate bundles. This means duplicated code and bloated app size

File: app/ancillaryFactory/src/config/codepushConfig.js (hunk 13 above)
Comment (tbergq) [reply app/ancillaryFactory/src/config/codepushConfig.js]: We already got this covered this file should be removed

File: app/ancillaries-native.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+// TODO: please check if it's still needed
+YellowBox.ignoreWarnings([
+  'Class GenericShare was not exported. Did you forget to use RCT_EXPORT_MODULE()',
+  'Class WhatsAppShare was not exported. Did you forget to use RCT_EXPORT_MODULE()',
+  'Class GooglePlusShare was not exported. Did you forget to use RCT_EXPORT_MODULE()',
+  'Class InstagramShare was not exported. Did you forget to use RCT_EXPORT_MODULE()',
+]);
+
+// Ancillaries
+AppRegistry.registerComponent('AncillaryFactory', () => AncillaryFactory);
```
Comment (tbergq) [inline app/ancillaries-native.js]: We should only register new modules in `native.js`

File: app/ancillaryFactory/package.json [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  "dependencies": {
+    "@kiwicom/mobile-config": "^0",
+    "@kiwicom/mobile-localization": "^0",
+    "@kiwicom/react-native-fast-track": "^0",
+    "@kiwicom/mobile-networking": "^0",
+    "@kiwicom/mobile-shared": "^0",
+    "react-native-code-push": "^5.5.2"
+  },
+  "devDependencies": {
+    "@kiwicom/mobile-playground": "^0"
```
Comment (tbergq) [inline app/ancillaryFactory/package.json]: I don't think you are using this dependency

File: app/ancillaryFactory/src/components/NoAncillaryMessage.js [hunk 16]
```diff
@@ -0,0 +1,16 @@
+// @flow strict
+
+import * as React from 'react';
+import { View } from 'react-native';
+import { Text } from '@kiwicom/mobile-shared';
+import { Translation } from '@kiwicom/mobile-localization';
+
+const NoAncillary = () => (
+  <View>
+    <Text>
```
Comment (tbergq) [inline app/ancillaryFactory/src/components/NoAncillaryMessage.js]: You actually don't need this text tag. The translation renders the default text tag. You just need this if you need to apply some styles.

File: app/native.js [hunk 17]
```diff
@@ -26,9 +25,6 @@ AppRegistry.registerComponent(
   () => SingleHotelStandalonePackage,
 );
 
-// Fast Track
-AppRegistry.registerComponent('FastTrackBanner', () => FastTrackBanner);
-
```
Comment (tbergq) [inline app/native.js]: This is the only place we should register components

File: ios/reactNativeApp.xcodeproj/xcshareddata/xcschemes/RNKiwiMobile.xcscheme [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+      </MacroExpansion>
+   </ProfileAction>
+   <AnalyzeAction
+      buildConfiguration = "Debug">
+   </AnalyzeAction>
+   <ArchiveAction
+      buildConfiguration = "Release"
+      revealArchiveInOrganizer = "YES">
+   </ArchiveAction>
+</Scheme>
```
Comment (tbergq) [inline ios/reactNativeApp.xcodeproj/xcshareddata/xcschemes/RNKiwiMobile.xcscheme]: What does this file do?

File: ios/reactNativeApp.xcodeproj/xcshareddata/xcschemes/reactNativeApp.xcscheme [hunk 19]
```diff
@@ -48,6 +48,20 @@
...
+            buildForArchiving = "YES"
+            buildForAnalyzing = "YES">
+            <BuildableReference
+               BuildableIdentifier = "primary"
+               BlueprintIdentifier = "27E931AF20AD96780053AE19"
+               BuildableName = "RNKiwiMobile.framework"
+               BlueprintName = "RNKiwiMobile"
+               ReferencedContainer = "container:reactNativeApp.xcodeproj">
+            </BuildableReference>
+         </BuildActionEntry>
```
Comment (tbergq) [inline ios/reactNativeApp.xcodeproj/xcshareddata/xcschemes/reactNativeApp.xcscheme]: What does these changes do?

File: package.json [hunk 20]
```diff
@@ -10,6 +10,7 @@
     "android": "yarn react-native run-android",
     "android-playground": "yarn react-native run-android --appFolder playground",
     "code-push": "node ./scripts/codePush.js",
+    "code-push-ancillaries": "yarn code-push --ancillaries",
```
Comment (tbergq) [inline package.json]: one code-push script is enough

File: packages/networking/package.json [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+  "main": "index.js",
+  "dependencies": {
+    "axios": "^0.18.0"
+  },
+  "peerDependencies": {
+    "react": "^16",
+    "react-native": "^0.57",
+    "react-test-renderer": "^16"
+  }
+}
```
Comment (tbergq) [inline packages/networking/package.json]: Please delete this package, we already got `@kiwicom/fetch`

**Comment by tbergq** [comment]
Also, CI is very unhappy, please fix those errors.

**Comment by tbergq** [comment]
Also, please rebase the master branch into your branch.

File: scripts/codePush.js (hunk 12 above)
Comment (SoHotSoup) [reply scripts/codePush.js]: @tbergq We have added this because we can't use `RNKiwiSharedBridge` and if we are going to use the same bundle with separated bridge instances:
1. Each bridge have a larger bundle to load (slower first render)
2. We have to plan updates and releases together. (`master` should be always production ready, which is not aligned with current workflow)
3. Other thing is that native team said that they don't like current architecture where they have to know to much RN details in their app. e.g. it doesn't make sense for native mobile app to have `codepush` as dependency or if they want to use some view they have to know about how our bridge is working.

So, if we add it like this as a separate bundle we are just adding 3MB to the native app. That can be even smaller if we are going to use `unbundle` as a bundle command which is code splitting for React native and it results in smaller bundles and faster loads.

Maybe at some point, we can join again together, but that would require changes in our architecture.

Comment (tbergq) [reply scripts/codePush.js]: Well, then I suggest you create a new repo if you can not use our setup

Comment (SoHotSoup) [reply scripts/codePush.js]: We can't create new repo because we don't want our native app to load multiple frameworks right? And we still need to use some of the features you've already made.
Try to think of this repo as a mobile framework, which it is, we just provide some views to our native app.
You are working on hotels we are working on ancillary views for MMB. Your views are the full screen which has different architecture requirements then our views which should be inline.

**Comment by vblazenka** [comment]
@tbergq Since you said this PR is too big, and I need to do graphql implementation is it okay to close this PR and I will open like 2-3 new ones? Depending on the size.

**Comment by tbergq** [comment]
Yes please, I think that is the way to go 😊
